olimex/teres_i/default.nix

```diff
@@ -0,0 +1,30 @@
+{ config, lib, pkgs, ...}:
+
+{
+  hardware.deviceTree = {
+    name = lib.mkDefault "allwinner/sun50i-a64-teres-i.dts";
+    enable = lib.mkDefault true;
+  };
+
+  boot = {
+    consoleLogLevel = lib.mkDefault 7;
+
+    kernelPackages = lib.mkDefault pkgs.linuxPackages_latest;
+    kernelParams = lib.mkDefault ["console=ttyS0,115200n8"];
+    extraModulePackages = lib.mkDefault [];
```

Comment:
This likely does nothing. See previous comment.

Reply:
[boot.kernelPackages defaults to `pkgs.linuxPackages` where we expect to use latest kernel](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/nixos/modules/system/boot/kernel.nix#L41)

`kernelParams` is important to provide the serial console, it was tested to not do that without the setting.

`extraModulePackages` is to ensure that it's empty by default

Reply:
There is a fundamental misunderstanding of how the modules system work in here.

`boot.kernelParams` ***with*** `lib.mkDefault` will not do what you expect. I suspect what you mean here is that you have tested `boot.kernelParams = ["console=ttyS0,115200n8"];` in your own config, then attempted to put it in here without validating that it was in use otherwise.

Here's a quick proof, this is evaluating a bare entirely unconfigured system, with your "configuration.nix" being only what is in the `configuration` arg.

```
~ $ nix repl --system aarch64-linux --arg configuration '{ lib, ... }: { boot.kernelParams = lib.mkDefault [ "THIS_IS_A_TEST" ]; }' '<nixpkgs/nixos>'

nix-repl> config.boot.kernelParams                              
[ "loglevel=4" ]
```

You will see that *it will never* end-up being used, as the moment *anything* sets `boot.kernelParams`, the priority is higher than the priority at `mkDefault`, and thus the whole list gets ignored.

Doing so for `boot.extraModulePackages` is also entirely unnecessary because NixOS modules system values need to be set to be used, and since it's always used, it is already set to a default value.

 - https://github.com/NixOS/nixpkgs/blob/81e8f48ebdecf07aab321182011b067aafc78896/nixos/modules/system/boot/kernel.nix#L164

Reply:
I tested both through my config and sourcing the fork, but i didn't notice different behavior

i will investigate and get back to you